### PR TITLE
constant fold 1-d numpy arrays

### LINKED ISSUE #1178: Fix constant folding for Tensor([3])
I believe now 1-D numpy arrays aren't being constant folded, they should be.

Like `Tensor(3)` works but `Tensor([3])` doesn't.

We should also add a test to prevent this regression in the future.

### PR DESCRIPTION
**Pull request by carsonRadtke**
Fixes #1178

### CODE REVIEW

**Comment by tinyb0t** [comment]
Changes made in `tinygrad/`:
```
------------------------------------------------------------
files                             insertions       deletions
------------------------------------------------------------
tinygrad/tensor.py                         4               4
------------------------------------------------------------
lines added in the tinygrad folder: 0
```

File: tinygrad/tensor.py [hunk 1]
```diff
@@ -58,13 +58,13 @@ def __init__(self, data:Union[int, float, list, tuple, LazyBuffer, np.ndarray],
...
 
-    if isinstance(data, (int, float)):
-      self.lazydata = LazyBuffer.loadop(LoadOps.CONST, tuple(), dtype or Tensor.default_type, device, data)
-      return
-
     if data.__class__ is list:
       data = np.array(data, dtype=(dtype or Tensor.default_type).np)
 
+    if isinstance(data, (int, float)) or data.__class__ is np.ndarray and data.size == 1:
+      self.lazydata = LazyBuffer.loadop(LoadOps.CONST, data.shape if data.__class__ is np.ndarray else tuple(), dtype or Tensor.default_type, device, data)
```
Comment (geohot) [inline tinygrad/tensor.py:65]: Err, I don't think you want the data to be a numpy array.

File: docs/abstractions.py [hunk 2]
```diff
@@ -120,7 +120,7 @@ class LoadOps(Enum):     EMPTY = auto(); RAND = auto(); CONST = auto(); FROM = a
 from tinygrad.ops import LazyOp, BinaryOps, LoadOps
 
 # the 2+3 from before
-result = Tensor([2]) + Tensor([3])
+result = Tensor([2, 3]) + Tensor([3, 4])
```
Comment (wozeparrot) [inline docs/abstractions.py:123]: ???

File: tinygrad/tensor.py [hunk 3]
```diff
@@ -58,17 +58,20 @@ def __init__(self, data:Union[int, float, list, tuple, LazyBuffer, np.ndarray],
...
       return
 
-    if isinstance(data, (int, float)):
-      self.lazydata = LazyBuffer.loadop(LoadOps.CONST, tuple(), dtype or Tensor.default_type, device, data)
-      return
-
     if data.__class__ is list:
       data = np.array(data, dtype=(dtype or Tensor.default_type).np)
 
+    if isinstance(data, (int, float)):
```
Comment (geohot) [inline tinygrad/tensor.py:64]: Leftover?

File: tinygrad/tensor.py [hunk 4]
```diff
@@ -58,17 +58,20 @@ def __init__(self, data:Union[int, float, list, tuple, LazyBuffer, np.ndarray],
...
+      self.lazydata = LazyBuffer.loadop(LoadOps.CONST, cast(np.ndarray, data).shape if data.__class__ is np.ndarray else tuple(), dtype or Tensor.default_type, device, data)
+      return
+
     if data.__class__ is np.ndarray:
       data = cast(np.ndarray, data)
+      dims, sz = data.ndim, data.size
       data = LazyBuffer.fromCPU(data)
       self.lazydata = data if data.device == device else LazyBuffer.loadop(LoadOps.FROM, data.shape, data.dtype, device, src=data)
+      if dims > 0 and sz == 1:
```
Comment (geohot) [inline tinygrad/tensor.py:73]: Why dims > 0?

File: tinygrad/tensor.py [hunk 5]
```diff
@@ -67,8 +67,11 @@ def __init__(self, data:Union[int, float, list, tuple, LazyBuffer, np.ndarray],
 
     if data.__class__ is np.ndarray:
       data = cast(np.ndarray, data)
+      sz = data.size
```
Comment (geohot) [inline tinygrad/tensor.py:70]: Why?

File: tinygrad/tensor.py [hunk 6]
```diff
@@ -69,6 +69,7 @@ def __init__(self, data:Union[int, float, list, tuple, LazyBuffer, np.ndarray],
       data = cast(np.ndarray, data)
       data = LazyBuffer.fromCPU(data)
       self.lazydata = data if data.device == device else LazyBuffer.loadop(LoadOps.FROM, data.shape, data.dtype, device, src=data)
+      if all(x == 1 for x in data.shape): self.lazydata = self.lazydata.const_like(data.toCPU().sum())
```
Comment (geohot) [inline tinygrad/tensor.py:72]: But why create the other one?

Comment (carsonRadtke) [reply tinygrad/tensor.py:72]: > Why is there a sum here?

`[0]` doesn't work if shape is `()`.

File: tinygrad/tensor.py [hunk 7]
```diff
@@ -67,10 +67,12 @@ def __init__(self, data:Union[int, float, list, tuple, LazyBuffer, np.ndarray],
 
     if data.__class__ is np.ndarray:
       data = cast(np.ndarray, data)
-      data = LazyBuffer.fromCPU(data)
-      self.lazydata = data if data.device == device else LazyBuffer.loadop(LoadOps.FROM, data.shape, data.dtype, device, src=data)
+      if data.shape == () or data.size == 1:
```
Comment (geohot) [inline tinygrad/tensor.py:70]: Do you need this or? Do you want to keep the shape?

File: test/test_tensor.py [hunk 8]
```diff
@@ -185,5 +186,16 @@ def test_element_size(self):
     for _, dtype in dtypes.fields().items():
       assert dtype.itemsize == Tensor.randn(3, dtype=dtype).element_size(), f"Tensor.element_size() not matching Tensor.dtype.itemsize for {dtype}"
 
+  def test_constant_fold(self):
+    a = Tensor(3)
+    b = Tensor([4])
+    c = Tensor([[5]])
+    assert a.lazydata.op.op == LoadOps.CONST
+    assert a.lazydata.op.arg == 3.0
+    assert b.lazydata.op.src[0].op.op == LoadOps.CONST
```
Comment (geohot) [inline test/test_tensor.py:195]: Why different?

**Comment by geohot** [comment]
I've now given a lot of feedback on this PR and it's still not ready to merge. Please think through all the implications of what you are writing and what my critiques will be before resubmitting.

**Pull request closed without merging**